Refactor process_goto_program so that all tools use common processing code

martin-cs opened the pull request:
`*_parse_optionst::process_goto_program` appears in `cbmc`, `goto-analyze` and `goto-diff` (and maybe the Java tools, @peterschrammel @hannes-steffenhagen-diffblue who is handling these at the moment?) with slight variations, bugs, inconsistencies and omissions.  This PR carefully makes them the same (so that we can `git-bisect` to find any issues this might introduce) and then pulls the common code out into a helper function.  This is another step towards a common format for goto-programs and stronger invariants.

This does not break tested functionality but it will have an impact on untested functionality.  Mostly, as far as I can see, it goes from 'silently doing the wrong thing' to 'maybe works, maybe breaks in an obvious way'.  i consider this a net win.

Commit:
Create a stub for a common process_goto_program

This should not alter observable behaviour.

Commit:
Add remove_asm to goto-analyzer to make program processing more consistent

This will change the behaviour of goto-analyzer.  I don't think it is
currently tested so I suspect it was unable to handle inline assembler.

Commit:
Enable the CPROVER library for goto-analyzer

This does change the functionality quite substantially but strangely it
only seems to affect one test, and that is syntactic rather than semantic.
As with many of the other changes in this PR, I suspect this is swapping
one lot of broken, untested functionality for another, slightly less broken
but still not working lot of functionality.

Commit:
Add 'malloc may fail' initialization to goto-diff

The tests aren't really sufficient to tell whether this has an effect.

Commit:
Add string abstraction to goto-analyzer and goto-diff

At the moment this code is dead and is only added to make the
process_goto_program steps more similar.  I will leave it to others
to decide if this is a useful thing to have and add the required command
line options and tests.  Should not affect the behaviour of either tool.

Commit:
Replace the use of cmdline with options in goto-analyzer and goto-diff

This uncovered a bug in goto-analyzer where the options for goto-check
are recognised but not actually parsed.

Commit:
Replace the parsing of goto_check options with the appropriate macro

This adds some functionality for options that were previously allowed
but silently ignored as they didn't make it from cmdline to goto_check.

Commit:
Add partial inlining as an optional part of process_goto_program

In CBMC and goto-diff this is dead, in goto-analyze it preserves the
existing behaviour.  It is not 100% clear to me that the existing behaviour
is useful or desirable but it is preserved as the point of this PR is not
changing behaviour unless necessary.

Commit:
Make the call to rewrite-union configurable

Rewrite union converts union member access to byte extracts.  This is
(apparently) good for CBMC and goto-diff but it makes analysis harder
for goto-analyzer.  So to make them consistent we add an option which
controls whether it is run or not.  This is fixed in each tool but it
takes us closer to common processing of goto programs.

Commit:
Add options that goto-check uses to the macros

goto-check makes use of the options assertions, assumptions,
assert-to-assume and retain-trivial.  These were handled to different
degrees and in different way in each of the tools.  This commit adds
these options to the set of options in the *_GOTO_CHECK macros.  This
reduces the amount of duplication and parallel maintenance.

Commit:
Enable goto_check in goto-analyzer

This was included in the options that goto-analyzer recognises and in its
--help output but they were all silently ignored.  This commit fixes that
so that they are supported, making it more consistent with CBMC (and less
inexplicable).

Commit:
goto-analyzer now adds the rounding mode to floating-point operations

Strangely (or perhaps concerningly) this doesn't seem to break any of
the regression tests.

Commit:
More string abstraction to be common processing of goto-programs

In the case of goto-analyzer and goto-diff this should be dead code.
In CBMC it reverses the order of string abstraction and non-det statics.
It is conceivable that these two interact but this way around is probably
better.

Commit:
Move the update functions earlier so that they are common between tools

remove_skip and various other passes compute the updates so moving this
should not effect the behaviour of the program.

martin-cs commented:
Commit by commit reviewing strongly encouraged.

tautschnig commented on src/goto-programs/process_goto_program.h:
```diff
@@ -0,0 +1,29 @@
+/// more complex types (vectors, complex, etc.) and constructs
+/// (returns, function pointers, etc.).
+/// This is can be used after initialize_goto_model but before
+/// analysis.  It is not madatory but is used by most tools.
```
Nit picking: s/madatory/mandatory/
But also: could/should this comment be moved towards the function documentation rather than the file's documentation?

martin-cs replied:
Fixed.

tautschnig commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -146,6 +146,9 @@ void goto_analyzer_parse_optionst::get_command_line_options(optionst &options)
 #endif
 
+  // all checks supported by goto_check
+  PARSE_OPTIONS_GOTO_CHECK(cmdline, options);
```
Ah! I had noticed at some point that I had to use `goto-instrument --*-check` before using `goto-analyzer`. Debugging `goto-analyzer` is still on my backlog...

martin-cs replied:
Yeah, a historical fuck up, esp. as the `GOTO_CHECK` options where given in `--help` and the options were parsed...

tautschnig commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -898,6 +900,11 @@ bool goto_analyzer_parse_optionst::process_goto_program(
+    mm_io(goto_model);
+
+    // instrument library preconditions
+    instrument_preconditions(goto_model);
```
Just noting that this isn't noted in the commit message. On a second thought, this change should probably go with the earlier commit that enabled library support in goto-analyzer.

martin-cs replied:
Fixed.

tautschnig commented on regression/goto-analyzer/approx-array-variable-const-fp/test.desc:
```diff
@@ -1,7 +1,7 @@
 main.c
 --show-goto-functions --pointer-check
-^Removing function pointers and virtual functions$
+^Removal of function pointers and virtual functions$
```
Perhaps put that change of the status message in a commit of its own so that the biggest change in this series does come without any observable impact on tests?

martin-cs replied:
Resolved

tautschnig reviewed (approved):
Thank you for this clean-up and doing it in such an easy-to-review way!

Commit:
Add memory mapped IO support to goto-analyzer

This does add functionality but it is of the "used to be silently ignored
but now sort of works or maybe breaks" kind.

Commit:
Change the text of the logging message so that it is the same as CBMC

This requires cosmetic changes to a lot of unit tests.

Commit:
Move the majority of process_goto_program out of the individual tools

This requires resolving a difference in the wording of a status message
that breaks a lot of tests for function pointer removal.

Commit:
Fix up the doxygen warnings

Commit:
Fix unit test that uses some, but not all, of CBMC's option parsing

martin-cs commented:
@hannes-steffenhagen-diffblue and / or @peterschrammel any chance of a review?  It looks big but commit by commit it is actually simple and should be easy to review quickly.

Commit:
Enabling goto_check in goto-analyzer adds 7 instructions to the test

peterschrammel commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -600,6 +581,10 @@ int goto_analyzer_parse_optionst::doit()
 
   goto_model = initialize_goto_model(cmdline.args, ui_message_handler, options);
 
+  // Perserve backwards compatability in processing
```
```suggestion
  // Preserve backwards compatibility in processing
```

martin-cs committed:
Fix typo in comments

Co-authored-by: Peter Schrammel <peterschrammel@users.noreply.github.com>
